### PR TITLE
cluster: reduce used memory by initialising http_async_client_ lazily

### PR DESCRIPTION
**Pull request by tpetkov-VMW**
Commit Message: Initialise http_async_client lazily inside ClusterManagerImpl
Additional Description: This saves memory when http_async_client isn't used
Risk Level: Low
Testing: unit test
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features:
[Optional Runtime guard:]
#19332
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by daixiang0** [comment]
Could you share details about this optimization?

**Comment by tpetkov-VMW** [comment]
> Could you share more details about this optimization? I do not get the benefit of lazy init, do I miss something important?

It's explained well in issue #19332 but the gist of it is that there are use cases for clusters which instantiate the Http::AsyncClientImpl async_http_client_  member, without using it. This is essentially wasted memory, multiplied by each thread and cluster.
I can try to give you concrete examples, if the ones in the issue aren't good enough.

**Comment by daixiang0** [comment]
Good explain!

Unfortunately, this is a bit complicated since the exact same code was posted in that issue.

I am not sure is it mandatory to need the original author to submit the work.

/cc @envoyproxy/envoy-maintainers

**Comment by tpetkov-VMW** [comment]
I think we can somehow co-author, but I can also ping him at least for verbal approval that it's okay.
He's on my team, I'm helping him out, so we will work it out for sure.

**Comment by daixiang0** [comment]
Good to hear that, ping @envoyproxy/envoy-maintainers to trigger CI.

**Comment by ppadevski** [comment]
I'm okay with not being listed as a co-author. Let's just try to merge as many of our patches as possible.

File: source/common/upstream/cluster_manager_impl.cc [hunk 1]
```diff
@@ -1104,7 +1104,15 @@ Host::CreateConnectionData ClusterManagerImpl::ThreadLocalClusterManagerImpl::Cl
 
 Http::AsyncClient&
 ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::httpAsyncClient() {
-  return http_async_client_;
+  if (!http_async_client_) {
```
Comment (jmarantz) [inline source/common/upstream/cluster_manager_impl.cc:1107]: nit: `http_async_client_ != nullptr`

**Review by jmarantz — COMMENTED** [review]
lgtm but I'll let someone who's worked more deeply in the data plane take a pass.

**Comment by alyssawilk** [comment]
Sorry googlers were out Friday but @KBaichoo should be able to take a look today!

File: source/common/upstream/cluster_manager_impl.h [hunk 2]
```diff
@@ -516,7 +516,7 @@ class ClusterManagerImpl : public ClusterManager,
       // Current active LB.
       LoadBalancerPtr lb_;
       ClusterInfoConstSharedPtr cluster_info_;
-      Http::AsyncClientImpl http_async_client_;
+      Http::AsyncClientPtr http_async_client_;
```
Comment (KBaichoo) [inline source/common/upstream/cluster_manager_impl.h:519]: Can you document that this is lazily created, and as such should probably rely on the encapsulation method ?

**Review by KBaichoo — COMMENTED** [review]
This seems reasonable to me, especially since http_async_client_ is well encapsulated. I can see the memory win if this isn't used.

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**